src/libsystemd/sd-device/sd-device.c

```diff
@@ -172,14 +172,12 @@ int device_set_syspath(sd_device *device, const char *_syspath, bool verify) {
                         return -ENODEV;
                 else if (r == -EINVAL) {
                         /* not a symlink */
-                        syspath = canonicalize_file_name(_syspath);
-                        if (!syspath) {
-                                if (errno == ENOENT)
-                                        /* the device does not exist (any more?) */
-                                        return -ENODEV;
-
-                                return log_debug_errno(errno, "sd-device: could not canonicalize '%s': %m", _syspath);
-                        }
+                        r = chase_symlinks(_syspath, NULL, 0, &syspath);
```

Comment:
Hmm I don't really grok the old btw, I think we should just call chase_symlinks() right-away instead of resolving the first step with readlink_and_canonicalize() and then the rest with chase_symlinks(). This appears unnecessarily complex still